### PR TITLE
[Validator] Checking a BIC along with an IBAN

### LINKED ISSUE #28166: [Validator] Validate BIC based on associated IBAN
**Description**  
A common use case is to save an IBAN with a BIC. Symfony provides validators for both properties but they do not communicate. 

We could add an optional propertyPath option (like http://symfony.com/doc/current/reference/constraints/GreaterThan.html#propertypath) on the BIC validator to check if the country code of the BIC is the same as the country code of the IBAN.

I can submit a PR if you want.

**Example**  
For instance a user can fill a French IBAN with a German BIC which will pass both validators but a failure could be expected somewhere as it does not make sense to fill a form/entity like this.

### PR DESCRIPTION
**Pull request by sylfabre**
| Q             | A
| ------------- | ---
| Branch?       | master
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | #28166 
| License       | MIT
| Doc PR        | https://github.com/symfony/symfony-docs/pull/10349

A BIC comes usually with an IBAN so it's better to check that they are associated. This PR provides an `iban` option to `Symfony\Component\Validator\Constraints\Bic` to check the BIC against an IBAN.

It also provides an `ibanPropertyPath` to retrieves the IBAN using the property accessor like with comparison constraints.

### CODE REVIEW

File: src/Symfony/Component/Validator/Constraints/Bic.php [hunk 1]
```diff
@@ -36,4 +39,23 @@ class Bic extends Constraint
     );
 
     public $message = 'This is not a valid Business Identifier Code (BIC).';
+
+    public $ibanMessage = 'This Business Identifier Code (BIC) is not associated with IBAN {{ iban }}.';
+
+    public $iban;
+
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Validator/Constraints/Bic.php]: I'd suggest to remove the blank lines between the added properties

File: src/Symfony/Component/Validator/Constraints/Bic.php [hunk 2]
```diff
@@ -36,4 +39,23 @@ class Bic extends Constraint
...
+    public $ibanPropertyPath;
+
+    public function __construct($options = null)
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Validator/Constraints/Bic.php]: missing "array" type hint?

File: src/Symfony/Component/Validator/Constraints/BicValidator.php [hunk 3]
```diff
@@ -81,5 +92,43 @@ public function validate($value, Constraint $constraint)
...
+
+            try {
+                $iban = $this->getPropertyAccessor()->getValue($object, $path);
+            } catch (NoSuchPropertyException $e) {
+                throw new ConstraintDefinitionException(sprintf('Invalid property path "%s" provided to "%s" constraint: %s', $path, \get_class($constraint), $e->getMessage()), 0, $e);
+            }
+        } else {
+            $iban = $constraint->iban;
+        }
+        if ($iban) {
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Validator/Constraints/BicValidator.php]: if (!$iban) => return early?

File: src/Symfony/Component/Validator/Constraints/BicValidator.php [hunk 4]
```diff
@@ -81,5 +92,43 @@ public function validate($value, Constraint $constraint)
...
+            $ibanCountryCode = substr($iban, 0, 2);
+            if (ctype_alpha($ibanCountryCode)) {
+                if (substr($canonicalize, 4, 2) !== $ibanCountryCode) {
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Validator/Constraints/BicValidator.php]: the two "if"s should be combined in one using `&&`

File: src/Symfony/Component/Validator/Constraints/BicValidator.php [hunk 5]
```diff
@@ -81,5 +92,43 @@ public function validate($value, Constraint $constraint)
...
+                    $this->context->buildViolation($constraint->ibanMessage)
+                        ->setParameter('{{ value }}', $this->formatValue($value))
+                        ->setParameter('{{ iban }}', $iban)
+                        ->setCode(Bic::INVALID_IBAN_COUNTRY_CODE_ERROR)
+                        ->addViolation();
+
+                    return;
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Validator/Constraints/BicValidator.php]: not needed

File: src/Symfony/Component/Validator/Tests/Constraints/BicValidatorTest.php [hunk 6]
```diff
@@ -36,6 +57,113 @@ public function testEmptyStringIsValid()
...
+        ));
+    }
+
+    public function testInvalidValuePath()
+    {
+        $constraint = new Bic(array('ibanPropertyPath' => 'foo'));
+
+        if (method_exists($this, 'expectException')) {
+            $this->expectException(ConstraintDefinitionException::class);
+            $this->expectExceptionMessage(sprintf('Invalid property path "foo" provided to "%s" constraint', \get_class($constraint)));
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Validator/Tests/Constraints/BicValidatorTest.php]: I'm wondering: what's the full message like? How much repetition does it contain, if any?

**Review by nicolas-grekas — COMMENTED** [review]
Thanks for working on this! Here are some random comments :)

**Comment by nicolas-grekas** [comment]
Could you improve the PR and commit title please also? (I'd suggest to squash commits while doing so)

File: src/Symfony/Component/Validator/Constraints/BicValidator.php (hunk 3 above)
Comment (sylfabre) [reply src/Symfony/Component/Validator/Constraints/BicValidator.php]: @nicolas-grekas no because the logic of validators is to return `null` on the first violation.

If tomorrow we add some more checks after this new code then we don't want to return early in this check.

File: src/Symfony/Component/Validator/Constraints/BicValidator.php (hunk 5 above)
Comment (sylfabre) [reply src/Symfony/Component/Validator/Constraints/BicValidator.php]: @nicolas-grekas I can remove it but I've done like the other violations raised above

File: src/Symfony/Component/Validator/Tests/Constraints/BicValidatorTest.php (hunk 6 above)
Comment (sylfabre) [reply src/Symfony/Component/Validator/Tests/Constraints/BicValidatorTest.php]: @nicolas-grekas I've copied-pasted the code from https://github.com/symfony/symfony/blob/master/src/Symfony/Component/Validator/Tests/Constraints/AbstractComparisonValidatorTestCase.php#L176

File: src/Symfony/Component/Validator/Constraints/BicValidator.php (hunk 3 above)
Comment (nicolas-grekas) [reply src/Symfony/Component/Validator/Constraints/BicValidator.php]: We shouldn't base current code on hypothetical future-based arguments. Here it would make the code clearer *now*.

File: src/Symfony/Component/Validator/Constraints/BicValidator.php (hunk 5 above)
Comment (nicolas-grekas) [reply src/Symfony/Component/Validator/Constraints/BicValidator.php]: Please remove it yes, the code elsewhere is unrelated.

File: src/Symfony/Component/Validator/Tests/Constraints/BicValidatorTest.php (hunk 6 above)
Comment (nicolas-grekas) [reply src/Symfony/Component/Validator/Tests/Constraints/BicValidatorTest.php]: I'd still be interested in the answer :)

Comment (sylfabre) [reply src/Symfony/Component/Validator/Tests/Constraints/BicValidatorTest.php]: @nicolas-grekas Invalid property path "foo" provided to "\Symfony\Component\Validator\Constraints\Bic" constraint

File: src/Symfony/Component/Validator/Constraints/BicValidator.php [hunk 7]
```diff
@@ -81,5 +92,38 @@ public function validate($value, Constraint $constraint)
...
+                try {
+                    $iban = $this->getPropertyAccessor()->getValue($object, $path);
+                } catch (NoSuchPropertyException $e) {
+                    throw new ConstraintDefinitionException(sprintf('Invalid property path "%s" provided to "%s" constraint: %s', $path, \get_class($constraint), $e->getMessage()), 0, $e);
+                }
+            }
+        } else {
+            $iban = $constraint->iban;
+        }
+        if (!$iban) {
```
Comment (ro0NL) [inline src/Symfony/Component/Validator/Constraints/BicValidator.php]: shouldnt we prefer `nulll === $iban`  instead?

File: src/Symfony/Component/Validator/Tests/Constraints/BicValidatorTest.php (hunk 6 above)
Comment (nicolas-grekas) [reply src/Symfony/Component/Validator/Tests/Constraints/BicValidatorTest.php]: Sorry it looks like I wasn't explicit enough: on the line above that throws the exception, the message ends with [...] constaint`: %s`, where %s is replaced by some `$e->getMessage()`. My question is about the value of this %s message: what does it look like in this test case for example?

File: src/Symfony/Component/Validator/Constraints/BicValidator.php (hunk 7 above)
Comment (sylfabre) [reply src/Symfony/Component/Validator/Constraints/BicValidator.php]: @nicolas-grekas tell me whatever you prefer!

File: src/Symfony/Component/Validator/Tests/Constraints/BicValidatorTest.php (hunk 6 above)
Comment (sylfabre) [reply src/Symfony/Component/Validator/Tests/Constraints/BicValidatorTest.php]: @nicolas-grekas Neither the property "foo" nor one of the methods "getFoo()", "foo()", "isFoo()", "hasFoo()", "__get()" exist and have public access in class "Symfony\Component\Validator\Tests\Constraints\BicComparisonTest_Class".

File: src/Symfony/Component/Validator/Constraints/BicValidator.php (hunk 7 above)
Comment (nicolas-grekas) [reply src/Symfony/Component/Validator/Constraints/BicValidator.php]: I'm fine with `!$iban` personally: the logic below also skips e.g. the empty string or other falsy values.

**Comment by nicolas-grekas** [comment]
Thanks, LGTM, except the test failure of course.

File: src/Symfony/Component/Validator/Tests/Constraints/BicValidatorTest.php (hunk 6 above)
Comment (sylfabre) [reply src/Symfony/Component/Validator/Tests/Constraints/BicValidatorTest.php]: Neither the property "foo" nor one of the methods "getFoo()", "foo()", "isFoo()", "hasFoo()", "__get()" exist and have public access in class "Symfony\Component\Validator\Tests\Constraints\BicComparisonTest_Class"

File: src/Symfony/Component/Validator/Constraints/BicValidator.php (hunk 7 above)
Comment (ro0NL) [reply src/Symfony/Component/Validator/Constraints/BicValidator.php]: right. different question then :)

shouldnt we expect a violation for arbitrary ibans?

Comment (nicolas-grekas) [reply src/Symfony/Component/Validator/Constraints/BicValidator.php]: you would get one by when adding a constraint on the iban field itself, isn't it enough?

**Review by nicolas-grekas — APPROVED** [review]
thanks

File: src/Symfony/Component/Validator/Constraints/BicValidator.php [hunk 8]
```diff
@@ -81,5 +92,38 @@ public function validate($value, Constraint $constraint)
...
+                try {
+                    $iban = $this->getPropertyAccessor()->getValue($object, $path);
+                } catch (NoSuchPropertyException $e) {
+                    throw new ConstraintDefinitionException(sprintf('Invalid property path "%s" provided to "%s" constraint: %s', $path, \get_class($constraint), $e->getMessage()), 0, $e);
+                }
+            }
+        } else {
+            $iban = $constraint->iban;
+        }
+        if (!isset($iban) || !$iban) {
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Validator/Constraints/BicValidator.php]: I'd suggest to rewrite the previous this way
```php
        $iban = $constraint->iban;
        $path = $constraint->ibanPropertyPath;
        if ($path && null !== $object = $this->context->getObject()) {
            try {
                $iban = $this->getPropertyAccessor()->getValue($object, $path);
            } catch (NoSuchPropertyException $e) {
                throw new ConstraintDefinitionException(sprintf('Invalid property path "%s" provided to "%s" constraint: %s', $path, \get_class($constraint), $e->getMessage()), 0, $e);
            }
        }
        if (!$iban) { 
```

**Review by javiereguiluz — APPROVED** [review]

File: src/Symfony/Component/Validator/Constraints/BicValidator.php [hunk 9]
```diff
@@ -91,5 +101,36 @@ public function validate($value, Constraint $constraint)
...
+                ->setParameter('{{ iban }}', $iban)
+                ->setCode(Bic::INVALID_IBAN_COUNTRY_CODE_ERROR)
+                ->addViolation();
+        }
+    }
+
+    private function getPropertyAccessor()
+    {
+        if (null === $this->propertyAccessor) {
+            $this->propertyAccessor = PropertyAccess::createPropertyAccessor();
```
[the remark this answers is not in the document]
Comment (nicolas-grekas) [reply src/Symfony/Component/Validator/Constraints/BicValidator.php]: correct! should leverage the new LogicException introduced in #28513

**Comment by nicolas-grekas** [comment]
Thanks @ro0NL for the comments, good catch!
@sylfabre OK to you with the comments? They all make sense to me (please rebase also.)

Status: needs work

File: src/Symfony/Component/Validator/Constraints/Bic.php [hunk 10]
```diff
@@ -36,4 +39,20 @@ class Bic extends Constraint
     );
 
     public $message = 'This is not a valid Business Identifier Code (BIC).';
+    public $ibanMessage = 'This Business Identifier Code (BIC) is not associated with IBAN {{ iban }}.';
```
[the remark this answers is not in the document]
Comment (sylfabre) [reply src/Symfony/Component/Validator/Constraints/Bic.php]: done

**Comment by sylfabre** [comment]
@nicolas-grekas @ro0NL I've rebased and updated the PR as per your comments.

I've also updated accordingly the code I used as source for my feature

File: src/Symfony/Component/Validator/Constraints/BicValidator.php [hunk 11]
```diff
@@ -23,6 +26,13 @@
  */
 class BicValidator extends ConstraintValidator
 {
+    private $propertyAccessor;
+
+    public function __construct(PropertyAccessor $propertyAccessor = null)
```
[the remark this answers is not in the document]
Comment (ro0NL) [reply src/Symfony/Component/Validator/Constraints/BicValidator.php]: looking at `AbstractComparisonValidator` i dont think we use property access as a service, so this seems fine :+1:

File: src/Symfony/Component/Validator/Tests/Constraints/AbstractComparisonValidatorTestCase.php [hunk 12]
```diff
@@ -16,26 +16,6 @@
 use Symfony\Component\Validator\Exception\ConstraintDefinitionException;
 use Symfony\Component\Validator\Test\ConstraintValidatorTestCase;
 
-class ComparisonTest_Class
```
Comment (ro0NL) [inline src/Symfony/Component/Validator/Tests/Constraints/AbstractComparisonValidatorTestCase.php]: ok, i see why you added the new class on top as well in `BicValidatorTest` :)

i still think we prefer classes at the bottom, but technically this change applies to 3.4 and is not really related to your PR.

Comment (sylfabre) [reply src/Symfony/Component/Validator/Tests/Constraints/AbstractComparisonValidatorTestCase.php]: @ro0NL should I remove my changes to these files then? and someone else makes another PR for this?

Comment (ro0NL) [reply src/Symfony/Component/Validator/Tests/Constraints/AbstractComparisonValidatorTestCase.php]: i think we should avoid this conflict with lower (maintained) branches yes

File: src/Symfony/Component/Validator/Constraints/BicValidator.php [hunk 13]
```diff
@@ -93,5 +104,39 @@ public function validate($value, Constraint $constraint)
...
+                ->setParameter('{{ iban }}', $iban)
+                ->setCode(Bic::INVALID_IBAN_COUNTRY_CODE_ERROR)
+                ->addViolation();
+        }
+    }
+
+    private function getPropertyAccessor(): PropertyAccessor
+    {
+        if (null === $this->propertyAccessor) {
+            if (!class_exists('Symfony\Component\PropertyAccess\PropertyAccess')) {
```
Comment (xabbuh) [inline src/Symfony/Component/Validator/Constraints/BicValidator.php]: If I don't miss anything, this cannot happen as the constraint would have thrown the exception before.

File: src/Symfony/Component/Validator/Constraints/Bic.php [hunk 14]
```diff
@@ -36,4 +39,20 @@ class Bic extends Constraint
...
+    public $iban;
+    public $ibanPropertyPath;
+
+    public function __construct(array $options = null)
+    {
+        if (isset($options['iban']) && isset($options['ibanPropertyPath'])) {
+            throw new ConstraintDefinitionException(sprintf('The "%s" constraint requires only one of the "iban" or "ibanPropertyPath" options to be set, not both.', \get_class($this)));
```
Comment (xabbuh) [inline src/Symfony/Component/Validator/Constraints/Bic.php]: I would reword this a bit:

> The "iban" and "ibanPropertyPath" of the Iban constraint cannot be used at the same time.

File: src/Symfony/Component/Validator/Constraints/Bic.php [hunk 15]
```diff
@@ -36,4 +39,20 @@ class Bic extends Constraint
...
+        }
+
+        if (isset($options['propertyPath']) && !class_exists(PropertyAccess::class)) {
+            throw new ConstraintDefinitionException(sprintf('The "%s" constraint requires the Symfony PropertyAccess component to use the "ibanPropertyPath" option.', \get_class($this)));
```
Comment (ro0NL) [inline src/Symfony/Component/Validator/Constraints/Bic.php]: `Symfony\Component\Validator\Exception\LogicException` here

File: src/Symfony/Component/Validator/Constraints/BicValidator.php (hunk 13 above)
Comment (ro0NL) [reply src/Symfony/Component/Validator/Constraints/BicValidator.php]: that would apply to all constraints where we move the check: #28644

Comment (sylfabre) [reply src/Symfony/Component/Validator/Constraints/BicValidator.php]: @nicolas-grekas should I then remove the exception here as it's already thrown in the constraint constructor?

File: src/Symfony/Component/Validator/Constraints/BicValidator.php (hunk 9 above)
[the remark this answers is not in the document]
Comment (sylfabre) [reply src/Symfony/Component/Validator/Constraints/BicValidator.php]: answered in previous comment

**Comment by sylfabre** [comment]
> @sylfabre up to address the remaining comments?

I've rebased the branch onto master and there is one question for you

File: src/Symfony/Component/Validator/Constraints/BicValidator.php (hunk 13 above)
Comment (sylfabre) [reply src/Symfony/Component/Validator/Constraints/BicValidator.php]: @nicolas-grekas I think this `getPropertyAccessor()` should be remove: we either use DI in the constructor to get the `PropertyAccess` component or use the static way to get it (like for Intl)

Comment (ro0NL) [reply src/Symfony/Component/Validator/Constraints/BicValidator.php]: the `BicValidator` has no `instanceof Bic` test (should be added though for consistency).

Technically this throw can still happen if called before `validate()` is called (unlikely, but possible from API pov)

I think it makes sense to keep the check; it's more easy to reason about, causes less confusion and makes no assumptions. And all constraints do it already, we can keep it even in 5.0. That's the shortest path towards consistency concerning accessing deps.

File: src/Symfony/Component/Validator/CHANGELOG.md [hunk 16]
```diff
@@ -13,6 +13,7 @@ CHANGELOG
  * deprecated using the `Bic`, `Country`, `Currency`, `Language` and `Locale` constraints without `symfony/intl`
  * deprecated using the `Email` constraint without `egulias/email-validator`
  * deprecated using the `Expression` constraint without `symfony/expression-language`
+ * validating a BIC based on a given IBAN
```
Comment (ro0NL) [inline src/Symfony/Component/Validator/CHANGELOG.md]: for discovery, id say:

- added options `iban` and `propertyPath` to `Bic` constraint

File: src/Symfony/Component/Validator/Constraints/Bic.php [hunk 17]
```diff
@@ -28,6 +30,7 @@ class Bic extends Constraint
     const INVALID_BANK_CODE_ERROR = '00559357-6170-4f29-aebd-d19330aa19cf';
     const INVALID_COUNTRY_CODE_ERROR = '1ce76f8d-3c1f-451c-9e62-fe9c3ed486ae';
     const INVALID_CASE_ERROR = '11884038-3312-4ae5-9d04-699f782130c7';
+    const INVALID_IBAN_COUNTRY_CODE_ERROR = '29a2c3bb-587b-4996-b6f5-53081364cea5';
```
Comment (ro0NL) [inline src/Symfony/Component/Validator/Constraints/Bic.php]: `public const`, but feel free to ignore. This is a CS discussion also (we added public for new consts before though)

File: src/Symfony/Component/Validator/Constraints/Bic.php [hunk 18]
```diff
@@ -38,14 +41,25 @@ class Bic extends Constraint
...
             // throw new LogicException(sprintf('The "symfony/intl" component is required to use the "%s" constraint.', __CLASS__));
             @trigger_error(sprintf('Using the "%s" constraint without the "symfony/intl" component installed is deprecated since Symfony 4.2.', __CLASS__), E_USER_DEPRECATED);
         }
 
+        if (isset($options['iban']) && isset($options['ibanPropertyPath'])) {
+            throw new ConstraintDefinitionException(sprintf('The "iban" and "ibanPropertyPath" of the Iban constraint cannot be used at the same time.', \get_class($this)));
+        }
+
+        if (isset($options['propertyPath']) && !class_exists(PropertyAccess::class)) {
+            throw new LogicException(sprintf('The "symfony/property-access" component is required to use the "%s" constraint with the "ibanPropertyPath" option.', \get_class($this)));
```
Comment (ro0NL) [inline src/Symfony/Component/Validator/Constraints/Bic.php]: `with the "propertyPath" option`

**Review by ro0NL — COMMENTED** [review]
LGTM to in general :+1: we need to decide if we add the `$constraint instanceof Bic` and throw (BC break-ish) or trigger deprecation.

File: src/Symfony/Component/Validator/Constraints/Bic.php (hunk 17 above)
Comment (nicolas-grekas) [reply src/Symfony/Component/Validator/Constraints/Bic.php]: let's keep consistency

**Comment by nicolas-grekas** [comment]
> we need to decide if we add the $constraint instanceof Bic and throw (BC break-ish) or trigger deprecation

easy to decide: no Bc breaks :)

File: src/Symfony/Component/Validator/Constraints/Bic.php (hunk 18 above)
Comment (sylfabre) [reply src/Symfony/Component/Validator/Constraints/Bic.php]: @ro0NL but the option really is `ibanPropertyPath`!

**Review by OskarStark — APPROVED** [review]
Anything todo here?

File: src/Symfony/Component/Validator/Constraints/Bic.php [hunk 19]
```diff
@@ -38,14 +41,25 @@ class Bic extends Constraint
...
-    public function __construct($options = null)
+    public function __construct(array $options = null)
     {
         if (!class_exists(Intl::class)) {
             // throw new LogicException(sprintf('The "symfony/intl" component is required to use the "%s" constraint.', __CLASS__));
             @trigger_error(sprintf('Using the "%s" constraint without the "symfony/intl" component installed is deprecated since Symfony 4.2.', __CLASS__), E_USER_DEPRECATED);
         }
 
+        if (isset($options['iban']) && isset($options['ibanPropertyPath'])) {
+            throw new ConstraintDefinitionException(sprintf('The "iban" and "ibanPropertyPath" of the Iban constraint cannot be used at the same time.', \get_class($this)));
```
Comment (ro0NL) [inline src/Symfony/Component/Validator/Constraints/Bic.php]: `... options of the Bic constraint ...`

**Review by ro0NL — COMMENTED** [review]
tests are red :)

File: src/Symfony/Component/Validator/Constraints/Bic.php [hunk 20]
```diff
@@ -38,14 +41,25 @@ class Bic extends Constraint
     );
 
     public $message = 'This is not a valid Business Identifier Code (BIC).';
+    public $ibanMessage = 'This Business Identifier Code (BIC) is not associated with IBAN {{ iban }}.';
+    public $iban;
+    public $ibanPropertyPath;
 
-    public function __construct($options = null)
+    public function __construct(array $options = null)
```
Comment (ro0NL) [inline src/Symfony/Component/Validator/Constraints/Bic.php]: not sure about this :)

**Comment by sylfabre** [comment]
@ro0NL tests are now green :)

File: src/Symfony/Component/Validator/Constraints/Bic.php (hunk 20 above)
Comment (sylfabre) [reply src/Symfony/Component/Validator/Constraints/Bic.php]: looks good for me as we expect an array.

Why aren't you sure about this?

Comment (xabbuh) [reply src/Symfony/Component/Validator/Constraints/Bic.php]: In this case, the change looks valid to me. Generally you can define a default option by overriding the `getDefaultOption()` method. But since that is not the case for the `Bic` constraint, it will already throw an exception when no array is passed.

File: src/Symfony/Component/Validator/CHANGELOG.md [hunk 21]
```diff
@@ -15,6 +15,7 @@ CHANGELOG
  * deprecated using the `Bic`, `Country`, `Currency`, `Language` and `Locale` constraints without `symfony/intl`
  * deprecated using the `Email` constraint without `egulias/email-validator`
  * deprecated using the `Expression` constraint without `symfony/expression-language`
+ * added options `iban` and `ibanPropertyPath` to Bic constraint
```
Comment (xabbuh) [inline src/Symfony/Component/Validator/CHANGELOG.md]: needs to be added to a new `4.3.0` section now as 4.2 is in feature freeze

File: src/Symfony/Component/Validator/Constraints/Bic.php [hunk 22]
```diff
@@ -38,14 +41,25 @@ class Bic extends Constraint
...
     {
         if (!class_exists(Intl::class)) {
             // throw new LogicException(sprintf('The "symfony/intl" component is required to use the "%s" constraint.', __CLASS__));
             @trigger_error(sprintf('Using the "%s" constraint without the "symfony/intl" component installed is deprecated since Symfony 4.2.', __CLASS__), E_USER_DEPRECATED);
         }
 
+        if (isset($options['iban']) && isset($options['ibanPropertyPath'])) {
+            throw new ConstraintDefinitionException(sprintf('The "iban" and "ibanPropertyPath" options of the Iban constraint cannot be used at the same time.', \get_class($this)));
```
Comment (xabbuh) [inline src/Symfony/Component/Validator/Constraints/Bic.php]: `self::class` instead of `get_class($this)`(same below)?

File: src/Symfony/Component/Validator/Constraints/BicValidator.php [hunk 23]
```diff
@@ -95,5 +106,39 @@ public function validate($value, Constraint $constraint)
...
```
Comment (xabbuh) [inline src/Symfony/Component/Validator/Constraints/BicValidator.php]: we should use the `class` constant here

Comment (sylfabre) [reply src/Symfony/Component/Validator/Constraints/BicValidator.php]: @xabbuh like `PropertyAccess::class` ?
I haven't tried but I think it will throw an exception if the class does not exist, right?

Comment (xabbuh) [reply src/Symfony/Component/Validator/Constraints/BicValidator.php]: yes like that, and using the `class` constant works even when the class is not present: https://3v4l.org/knSo3

File: src/Symfony/Component/Validator/Constraints/Bic.php (hunk 20 above)
Comment (ro0NL) [reply src/Symfony/Component/Validator/Constraints/Bic.php]: exactly, but now we'll never reach the exception at https://github.com/symfony/symfony/blob/491bbe741dc0762eeda59ceb7d441a520af8d878/src/Symfony/Component/Validator/Constraint.php#L136


so basically all constraints will throw this exception, except here it will cause a type error: https://3v4l.org/WjMuY

File: src/Symfony/Component/Validator/Constraints/BicValidator.php (hunk 23 above)
Comment (sylfabre) [reply src/Symfony/Component/Validator/Constraints/BicValidator.php]: @xabbuh thanks!

**Comment by nicolas-grekas** [comment]
Thank you @sylfabre.
